yarpl/include/yarpl/flowable/EmitterFlowable.h

```diff
@@ -26,25 +26,20 @@ class EmiterBase : public virtual Refcounted {
  * This is synchronous: the emit calls are triggered within the context
  * of a request(n) call.
  */
-template<typename T>
-class EmiterSubscription : private Subscription, private Subscriber<T> {
+template <typename T>
+class EmiterSubscription : public virtual Subscription,
+                           public virtual Subscriber<T> {
   constexpr static auto kCanceled = credits::kCanceled;
   constexpr static auto kNoFlowControl = credits::kNoFlowControl;
 
  public:
   EmiterSubscription(
-      Reference <EmiterBase<T>> emiter,
-      Reference <Subscriber<T>> subscriber)
-      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
-    // We expect to be heap-allocated; until this subscription finishes
-    // (is canceled; completes; error's out), hold a reference so we are
-    // not deallocated (by the subscriber).
-    Refcounted::incRef(*this);
-    subscriber_->onSubscribe(get_ref<Subscription>(this));
-  }
+      Reference<EmiterBase<T>> emiter,
+      Reference<Subscriber<T>> subscriber)
+      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {}
 
-  virtual ~EmiterSubscription() {
-    subscriber_.reset();
+  void init() {
+    subscriber_->onSubscribe(get_ref<Subscription>(this));
```

Comment:
Oh I see what's happening - similar problem to what was going on in `process`. There's a refcount of 0 on entry to the ctor, then we pass a reference to ourselves to `subscriber_->onSubscribe(get_ref(this))`, which doesn't keep a reference to `this` in some cases past the `onSubscribe` call, and once the temporary `get_ref` passed to `onSubscribe` gets destroyed, we call our dtor (while we're still in the ctor). 

Visually, if the onSubscribe is kept in the ctor: 
```cpp
  EmiterSubscription(
      Reference<EmiterBase<T>> emiter,
      Reference<Subscriber<T>> subscriber)
      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
        // count() == 0
        auto ref = get_ref<Subscription>(this);
        // count() == 1
        subscriber_->onSubscribe(ref);
        // count() == 0, dtor called while we're in the ctor
      }
```

So the solution here is to defer calling `onSubscribe` until we're out of the ctor for EmitterSubscription (by moving it into an `init()` method).